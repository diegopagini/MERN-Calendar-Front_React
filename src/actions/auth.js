/** @format */
import { fetchSinToken, fetchConToken } from '../helpers/fetch';
import { types } from '../types/types';
import Swal from 'sweetalert2';

export const startLogin = (email, password) => {
	return async (dispatch) => {
		const resp = await fetchSinToken('auth', { email, password }, 'POST');
		const body = await resp.json();

		if (body.ok) {
			localStorage.setItem('token', body.token);
			localStorage.setItem('token-init-date', new Date().getTime());

			dispatch(
				login({
					uid: body.uid,
					name: body.name,
				})
			);
		} else {
			Swal.fire('Error', body.msg, 'error');
		}
	};
};

export const startRegister = (email, password, name) => {
	return async (dispatch) => {
		const resp = await fetchSinToken(
			'auth/new',
			{ email, password, name },
			'POST'
		);
		const body = await resp.json();

		if (body.ok) {
			localStorage.setItem('token', body.token);
			localStorage.setItem('token-init-date', new Date().getTime());

			dispatch(
				login({
					uid: body.uid,
					name: body.name,
				})
			);
		} else {
			Swal.fire('Error', body.msg, 'error');
		}
	};
};

export const startChecking = () => {
	return async (dispatch) => {
		const resp = await fetchConToken('auth/renew');
		const body = await resp.json();

		if (body.ok) {
			localStorage.setItem('token', body.token);
			localStorage.setItem('token-init-date', new Date().getTime());

			dispatch(
				login({
					uid: body.uid,
					name: body.name,
				})
			);
		} else {
			dispatch(checkingFinish());
		}
	};
};

const login = (user) => ({
	type: types.authLogin,
	payload: user,
});

const checkingFinish = () => ({
	type: types.authCheckingFinish,
});

export const startLogout = () => {
	return (dispatch) => {
		localStorage.clear();
		dispatch(eventLogout());
		dispatch(logout());
	};
};

const logout = () => ({
	type: types.authLogout,
});

export const eventLogout = () => ({
	type: types.eventLogout,
});
