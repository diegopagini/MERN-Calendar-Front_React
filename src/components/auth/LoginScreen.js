/** @format */

import React from 'react';
import { useDispatch } from 'react-redux';
import { useForm } from '../../hooks/useForm';
import { startLogin, startRegister } from '../../actions/auth';
import Swal from 'sweetalert2';
import './login.css';

export const LoginScreen = () => {
	const [formLoginValues, handleLoginInputChange] = useForm({
		lEmail: '',
		lPassword: '',
	});

	const { lEmail, lPassword } = formLoginValues;

	const [formRegisterValues, handleRegisterInputChange] = useForm({
		rEmail: '',
		rName: '',
		rPassword1: '',
		rPassword2: '',
	});

	const { rName, rEmail, rPassword1, rPassword2 } = formRegisterValues;

	const dispatch = useDispatch();

	const handleLogin = (e) => {
		e.preventDefault();
		dispatch(startLogin(lEmail, lPassword));
	};

	const handleRegister = (e) => {
		e.preventDefault();
		if (rPassword1 !== rPassword2) {
			Swal.fire('Error', 'Las contrase単as deben de ser iguales', 'error');
		}
		dispatch(startRegister(rEmail, rPassword1, rName));
	};

	return (
		<div className='container login-container'>
			<div className='row'>
				<div className='col-md-6 login-form-1'>
					<h3>Ingreso</h3>
					<form onSubmit={handleLogin}>
						<div className='form-group'>
							<input
								type='text'
								className='form-control'
								placeholder='Correo'
								name='lEmail'
								value={lEmail}
								onChange={handleLoginInputChange}
							/>
						</div>
						<div className='form-group'>
							<input
								type='password'
								className='form-control'
								placeholder='Contrase単a'
								name='lPassword'
								value={lPassword}
								onChange={handleLoginInputChange}
							/>
						</div>
						<div className='form-group'>
							<input type='submit' className='btnSubmit' value='Login' />
						</div>
					</form>
				</div>

				<div className='col-md-6 login-form-2'>
					<h3>Registro</h3>
					<form onSubmit={handleRegister}>
						<div className='form-group'>
							<input
								type='text'
								className='form-control'
								placeholder='Nombre'
								name='rName'
								value={rName}
								onChange={handleRegisterInputChange}
							/>
						</div>
						<div className='form-group'>
							<input
								type='email'
								className='form-control'
								placeholder='Correo'
								name='rEmail'
								value={rEmail}
								onChange={handleRegisterInputChange}
							/>
						</div>
						<div className='form-group'>
							<input
								type='password'
								className='form-control'
								placeholder='Contrase単a'
								name='rPassword1'
								value={rPassword1}
								onChange={handleRegisterInputChange}
							/>
						</div>

						<div className='form-group'>
							<input
								type='password'
								className='form-control'
								placeholder='Repita la contrase単a'
								name='rPassword2'
								value={rPassword2}
								onChange={handleRegisterInputChange}
							/>
						</div>

						<div className='form-group'>
							<input type='submit' className='btnSubmit' value='Crear cuenta' />
						</div>
					</form>
				</div>
			</div>
		</div>
	);
};
