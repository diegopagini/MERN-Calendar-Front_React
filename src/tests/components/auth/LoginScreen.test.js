/** @format */

import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';

import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import Swal from 'sweetalert2';

import '@testing-library/jest-dom';
import { LoginScreen } from '../../../components/auth/LoginScreen';
import { startLogin, startRegister } from '../../../actions/auth';

jest.mock('../../../actions/auth', () => ({
	startLogin: jest.fn(),
	startRegister: jest.fn(),
}));

jest.mock('sweetalert2', () => ({
	fire: jest.fn(),
}));

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {};
const store = mockStore(initState);
store.dispatch = jest.fn();

const wrapper = mount(
	<Provider store={store}>
		<LoginScreen />
	</Provider>
);

describe('Pruebas en <LoginScreen />', () => {
	beforeEach(() => {
		jest.clearAllMocks();
	});

	test('debe mostrarse correctamente', () => {
		expect(wrapper).toMatchSnapshot();
	});

	test('debe de llamar el dispatch del login', () => {
		wrapper.find('input[name="lEmail"]').simulate('change', {
			target: {
				name: 'lEmail',
				value: 'juan@gmail.com',
			},
		});

		wrapper.find('input[name="lPassword"]').simulate('change', {
			target: {
				name: 'lPassword',
				value: '123456',
			},
		});

		wrapper.find('form').at(0).prop('onSubmit')({
			preventDefault() {},
		});

		expect(startLogin).toHaveBeenCalledWith('juan@gmail.com', '123456');
	});

	test('No hay registro si las contraseñas son diferentes', () => {
		wrapper.find('input[name="rPassword1"]').simulate('change', {
			target: {
				name: 'rPassword1',
				value: '123456',
			},
		});

		wrapper.find('input[name="rPassword2"]').simulate('change', {
			target: {
				name: 'rPassword2',
				value: '1234567',
			},
		});

		wrapper.find('form').at(1).prop('onSubmit')({
			preventDefault() {},
		});

		expect(startRegister).not.toHaveBeenCalled();
		expect(Swal.fire).toHaveBeenCalledWith(
			'Error',
			'Las contraseñas deben de ser iguales',
			'error'
		);
	});

	test('Registro con contraseñas iguales', () => {
		wrapper.find('input[name="rPassword1"]').simulate('change', {
			target: {
				name: 'rPassword1',
				value: 'hola mundo',
			},
		});

		wrapper.find('input[name="rPassword2"]').simulate('change', {
			target: {
				name: 'rPassword2',
				value: 'hola mundo',
			},
		});

		wrapper.find('form').at(1).prop('onSubmit')({
			preventDefault() {},
		});

		expect(Swal.fire).not.toHaveBeenCalled();
		expect(startRegister).toHaveBeenCalledWith(
			'nando@gmail.com',
			'hola mundo',
			'Nando'
		);
	});
});
