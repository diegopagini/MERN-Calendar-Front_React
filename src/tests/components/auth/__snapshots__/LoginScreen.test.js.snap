// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <LoginScreen /> debe mostrarse correctamente 1`] = `
<div
  className="container login-container"
>
  <div
    className="row"
  >
    <div
      className="col-md-6 login-form-1"
    >
      <h3>
        Ingreso
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="lEmail"
            onChange={[Function]}
            placeholder="Correo"
            type="text"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="lPassword"
            onChange={[Function]}
            placeholder="Contraseña"
            type="password"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="btnSubmit"
            type="submit"
            value="Login"
          />
        </div>
      </form>
    </div>
    <div
      className="col-md-6 login-form-2"
    >
      <h3>
        Registro
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="rName"
            onChange={[Function]}
            placeholder="Nombre"
            type="text"
            value="Nando"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="rEmail"
            onChange={[Function]}
            placeholder="Correo"
            type="email"
            value="nando@gmail.com"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="rPassword1"
            onChange={[Function]}
            placeholder="Contraseña"
            type="password"
            value="123456"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="rPassword2"
            onChange={[Function]}
            placeholder="Repita la contraseña"
            type="password"
            value="123456"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="btnSubmit"
            type="submit"
            value="Crear cuenta"
          />
        </div>
      </form>
    </div>
  </div>
</div>
`;
